### DISCUSSION TITLE
How to add canonical url dynamically from Route (tools/[tools]) next js

### DISCUSSION THREAD

**Discussion by rehamza1** [vercel/next.js#43596] [Help]
I added the canonical url dynamically by using useRouter and I am using drop down menu on frontend to change route key and I am using the simple next js only
 
1. I used this "router?.query?.tools" to get dynamically key 
and browser console show me the right key and also updated canonical url in head in (Elements = inspect chrome tool)

2. But project terminal console not showing the data if I use this router?.query?.tools it's show **undefined** and 
if I use router.asPath it's show  **/tools/[tools]** and also when I open (Sources = inspect chrome tool) it's shows me the default url in canonical not dynamicall updated url

I have question regarding SEO in 1. case it's working fine and shows me the canonical url n head in (Elements = inspect chrome tool) but SEO tools provide me error regarding canonical url So how I resolve this error and how to update canonical url dynamically

**1 tools/[tools] component**

`function Tools() {
  const router = useRouter();
  const [subUrlKey, setSubUrlKey] = useState('dmarc');
  console.log('Url keys value', router?.query?.tools, subUrlKey, router.asPath);

  return (
    <PageLayout pageKey="tools" pageSubKey={router?.query?.tools || 'dmarc'}>
      <Container maxWidth="lg" sx={{ pt: 2, pb: 6 }}>
        <ToolsPageView setSubUrlKey={setSubUrlKey} />
      </Container>
    </PageLayout>
  );
}

export default Tools;`

**2 Page Layout component**

`function PageLayout(props) {
  const router = useRouter();
  const { pageKey, children } = props || {};
  console.log(
    'Url keys value 22',
    pageKey,
    props?.pageSubKey,
    router?.query?.tools,
    router.asPath
  );

  return (
    <>
      <PageSEO pageKey={pageKey} pageSubKey={props?.pageSubKey} />

      <SnackbarProvider>
        <Header />
        {children}
        <Footer pageKey={pageKey} />
      </SnackbarProvider>
    </>
  );
}

export default PageLayout;`

**3 SEO component**

`function PageSEO(props) {
  const router = useRouter();
  const { pageKey } = props || {};
  if (!pageKey || !SEO_DATA || !SEO_DATA[pageKey]) return;

  const pageMetaInfo = SEO_DATA[pageKey];
  let { url } = pageMetaInfo;
  if (props?.pageSubKey) {
    url = `${url}/${props?.pageSubKey}`;
  }

  console.log(
    'Url keys value 3',
    pageKey,
    props?.pageSubKey,
    router?.query?.tools,
    router.asPath
  );

  return (
    <NextSeo
      title={pageMetaInfo.title}
      description={pageMetaInfo.description}
      canonical={url}
      openGraph={{
        type: 'website',
        url,
        title: pageMetaInfo.title,
        description: pageMetaInfo.description,
        images: [
          {
            width: 850,
            height: 600,
            url: pageMetaInfo.cover,
            alt: pageMetaInfo.title
          }
        ]
      }}
    />
  );
}

export default PageSEO;`

**Comment by icyJoseph** [comment]
> I added the canonical url dynamically by using useRouter and I am using drop down menu on frontend to change route key and I am using the simple next js only

That's looks already wrong in terms of naive SEO. Are you doing this in use-effect? Those run only in the client, so the initial HTML, seen by Bots or SEO won't see it. Some SEO tools do run JS though, but it is good to not make any assumptions.

Router query during pre-rendering is empty, unless the values are actually available via data fetching functions. Could you show a bit of code, or pseudo-code snippets of your pages?

**Reply by rehamza1** [reply]
Thanks for the response and I updated the discussion description you can see the code there

no I am not using useEffect but route is dynamically updated on frontend using dropdown and it's shows me the right key on browser console plus canonical url in Elements -> head 
but I am using **semrush** SEO tool it's show me canonical error and also it's not show in the Sources = inspect chrome tool
